## Data Manipulation and Analysis in Python


### Introduction
In this project, i perform  data manipulation and analysis on the Titanic Dataset from Kaggle (https://www.kaggle.com/competitions/titanic) using Python. I load the dataset, clean the data, and provide an overview of passengers who survived the Titanic shipwreck of 1912 by exploring variables like passenger socio-economic class, age, gender, and fare.

### Data Loading
Import required libraries and Load the Titanic dataset into a pandas Dataframe

In [1]:
import pandas as pd

### Read CSV file (dataset) into a pandas DataFrame

In [3]:
titanic_df = pd.read_csv(r"C:\Users\ext.carmen.salazar\OneDrive - DSV\Desktop\titanic_dataset.csv")

### Exploring the dataset
Use 'titanic_df.head() to view the first few rows of the dataframe and get an overview od the data'

In [4]:
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

View total number of rows in dataframe (total number of observations in dataset) to help make sense of the dataset size and sample size


In [5]:
total_observations = titanic_df.shape[0]
print("Total number of observations:", total_observations)

Total number of observations: 891


### Data Cleaning
Perform data cleaning procedures to address missing values, outliers, and inconsistencies.

Number of non-missing values in each column of the dataframe

In [6]:
column_counts = titanic_df.count()
print("Column counts:")
print(column_counts)


Column counts:
PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64


Total number of missing values in each column

In [7]:

missing_values = titanic_df.isnull().sum()
print("Missing values per column:")
print(missing_values)

Missing values per column:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


Fill in the missing values in 'Age' column with the mean age of available data because it perserves the overall distribution and central tendency of age data. The assumption is that missing data is at random.

In [8]:
mean_age = titanic_df['Age'].mean()
titanic_df['Age'].fillna(mean_age, inplace=True)

Handle missing values in 'Cabin' column. This column has a significant number of missing values, so creating a new binary feature 'HasCabin' allows for capturing whether a passenger had a cabin value or not. The absence of 'Cabin' data may be indicative of something meaningful.

In [9]:
titanic_df['HasCabin'] = titanic_df['Cabin'].notnull().astype(int)

Handle missing values in 'Embarked' column. Filling in only 2 missing values with the most frequent value to maintain overall distribution of embarkation ports. The assumption is that missing data is at random.

In [10]:
mode_embarked = titanic_df['Embarked'].mode()[0]
titanic_df['Embarked'].fillna(mode_embarked, inplace=True)

Check for and count duplicates in dataframe

In [11]:
# check for duplicates in dataframe
duplicates = titanic_df.duplicated()

# count the number of duplicates
number_duplicates = duplicates.sum()

print("Number of duplicate rows:", number_duplicates)

Number of duplicate rows: 0


Retrieve the data types of each column to see if i need to convert data types for proper analysis. It looks like assigned types align with expected type for each column in the dataset.

In [12]:
data_types = titanic_df.dtypes
print("Data types of each column:")
print(data_types)

Data types of each column:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
HasCabin         int32
dtype: object


### Data Transformation

#### What is the average age of passengers?
The average age of passengers on the Titanic (29.7) provides an understanding of the age distribution.

In [13]:
mean_age = titanic_df['Age'].mean()
print("Average age of passengers:", mean_age)

Average age of passengers: 29.69911764705882


#### What is the minimum and maximum age of passengers?

 The minimum and maximum provides an age range of passengers. The minimum age is 0.42 years (about 5 months) and the maximum age is 80 years. 

In [14]:
min_age = titanic_df['Age'].min()
max_age = titanic_df['Age'].max()

print("Minimum age of passengers:", min_age)
print("Maximum age of passengers:", max_age)

Minimum age of passengers: 0.42
Maximum age of passengers: 80.0


Calculate average age of survivors and non-survivors

In [16]:
avg_age_survived = titanic_df[titanic_df['Survived'] == 1]['Age'].mean()
avg_age_not_survived = titanic_df[titanic_df['Survived'] == 0]['Age'].mean()

print("Average age of survivors:", avg_age_survived)
print("Average age of non-survivors:", avg_age_not_survived)

Average age of survivors: 28.549778121775027
Average age of non-survivors: 30.415099646415943


#### What is the median fare paid by the passengers?
This demonstrates that half of the passengers paid less than $14.45 and the other half paid more than $14.45. This provides a central tendency that is less affected by extreme fare values.

In [17]:
median_fare = titanic_df['Fare'].median()
print("Median fare paid by passengers:", median_fare)

Median fare paid by passengers: 14.4542


Calculate average fare of survivors and non-survivors

In [18]:
avg_fare_survived = titanic_df[titanic_df['Survived'] == 1]['Fare'].mean()
avg_fare_not_survived = titanic_df[titanic_df['Survived'] == 0]['Fare'].mean()

print("Average fare of survivors:", avg_fare_survived)
print("Average fare of non-survivors:", avg_fare_not_survived)

Average fare of survivors: 48.39540760233918
Average fare of non-survivors: 22.117886885245902


#### What is the fare distribution and spread of prices paid?
Calculating the quartiles (25th, 50th, and 75th percentile) of "Fare" column. After calculating, we can see that 25th percentile of the fare variable is $7.91, the 75th percentile of the fare variables is $31, meaning that 25% of passengers paid less than $7.91, while 25% paid more than $31. 

In [17]:
quartiles = titanic_df['Fare'].quantile([0.25, 0.5, 0.75])
print("Quartiles of fare:")
print(quartiles)

Quartiles of fare:
0.25     7.9104
0.50    14.4542
0.75    31.0000
Name: Fare, dtype: float64


#### What is the most common socio-economic class among the passengers?
Mode: most frequent value(s) in passenger class variable to determine most common socio-economic class among the passengers

In [18]:
mode_passenger_class = titanic_df['Pclass'].mode()
print("Mode of passenger class:")
print(mode_passenger_class)

Mode of passenger class:
0    3
Name: Pclass, dtype: int64


In [19]:
# we can see that the column has multuple modes (0 & 3), to access most common passenger class, we use index at 0

common_passenger_class = mode_passenger_class[0]
print("most common passenger class:", common_passenger_class)

most common passenger class: 3


Calculate most common socio-economic class among survivors and non-survivors

In [19]:
pclass_survived = titanic_df[titanic_df['Survived'] == 1]['Pclass'].mode()[0]
pclass_not_survived = titanic_df[titanic_df['Survived'] == 0]['Pclass'].mode()[0]

print("Most common socio-economic class among survivors:", pclass_survived)
print("Most common socio-economic class among non-survivors:", pclass_not_survived)

Most common socio-economic class among survivors: 1
Most common socio-economic class among non-survivors: 3


#### Did women or men die more in the Titanic dataset?

Grouping data by 'Sex' column and calculating count of survivors and non-survivors within each group

In [21]:
survival_cnts = titanic_df.groupby('Sex')['Survived'].value_counts()

Number count of survivors and non-survivors for women and men

In [22]:
women_survived = survival_cnts.loc[('female', 1)]
women_not_survived = survival_cnts.loc[('female', 0)]

men_survived = survival_cnts.loc[('male', 1)]
men_not_survived = survival_cnts.loc[('male', 0)]

Calculate total number of women and men in dataset

In [23]:
total_women = women_survived + women_not_survived
total_men = men_survived + men_not_survived

Calculate percentage of survivors and non-survivors for women and men

In [24]:
women_survial_rate = women_survived / total_women * 100
men_survial_rate = men_survived / total_men * 100

Compare counts and survival rates to determine whether women or men had higher survival rate

In [25]:
print("Women Survived: ", women_survived)
print("Women not Survived: ", women_not_survived)

print("Men Survived: ", men_survived)
print("Men not Survived: ", men_not_survived)

print("Survival rate for women: ", women_survial_rate)
print("Survival rate for men: ", men_survial_rate)

Women Survived:  233
Women not Survived:  81
Men Survived:  109
Men not Survived:  468
Survival rate for women:  74.20382165605095
Survival rate for men:  18.890814558058924


### Interpretation and insights

The objective of this project was to determine factors that influenced the likelihood of survival among passengers. By analyzing passenger data such as age, gender, socio-economic class, and fare, i aim to understand which characteristics were associated with a higher probability of survival.

The results suggest that factors such as age, gender, socio-economic class, and fare influenced the probability of survival on the Titanic. Younger passengers, those with higher fares or from higher socio-economic classes (particularly class 1), and women had a higher likelihood of surviving the disaster.

Average age of survivors: 28.5
Average age of non-survivors: 30.4

Average fare of survivors: $48.4
Average fare of non-survivors: $22.1

Most common socio-economic class among survivors: 1 (Upper)
Most common socio-economic class among non-survivors: 3 (Lower)

Women Survived: 233
Men Survived:  109

Women not Survived: 81
Men not Survived:  468

Survival rate for women:  %74.2
Survival rate for men:  %18.9